File: app/src/main/java/in/completecourse/SplashActivity.java
package in.completecourse;

import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;

import com.google.android.gms.ads.MobileAds;
import com.google.firebase.analytics.FirebaseAnalytics;
import com.google.firebase.firestore.FirebaseFirestore;

import in.completecourse.helper.PrefManager;

public class SplashActivity extends AppCompatActivity {
    private static String versionCodeApp;
    private static String versionCode;
    private FirebaseFirestore db;


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        // Making notification bar transparent
        if (Build.VERSION.SDK_INT >= 21) {
            getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_STABLE | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN);
        }
        changeStatusBarColor();
        setContentView(R.layout.activity_splash);

        db = FirebaseFirestore.getInstance();

        //obtain the firebase analytics instance
        FirebaseAnalytics mFirebaseAnalytics = FirebaseAnalytics.getInstance(this);

        //Sample AdMob app ID: ca-app-pub-3940256099942544~3347511713
        MobileAds.initialize(this, SplashActivity.this.getResources().getString(R.string.admob_app_id));

        int versionCode = BuildConfig.VERSION_CODE;
        versionCodeApp = String.valueOf(versionCode);

        if (isNetworkAvailable()) {
            getVersionCode();
        }else {
            Toast.makeText(SplashActivity.this, "Please Check your Internet Connection", Toast.LENGTH_LONG).show();
        }
    }
    /**
     * Making notification bar transparent
     */
    private void changeStatusBarColor() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            Window window = getWindow();
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
            window.setStatusBarColor(Color.TRANSPARENT);
        }
    }

    static {
        AppCompatDelegate.setCompatVectorFromResourcesEnabled(true);
    }


    /**
     * Checks if there is Internet accessible.
     * Based on a stackoverflow snippet
     *
     * @return True if there is Internet. False if not.
     */
    private boolean isNetworkAvailable() {
        NetworkInfo activeNetworkInfo = null;
        ConnectivityManager connectivityManager
                = (ConnectivityManager) this.getSystemService(Context.CONNECTIVITY_SERVICE);
        if (connectivityManager!=null){
            activeNetworkInfo = connectivityManager.getActiveNetworkInfo();
        }
        return activeNetworkInfo != null && activeNetworkInfo.isConnected();
    }


    private void getVersionCode(){
        db.collection("flags").document("version_code").get().addOnSuccessListener(documentSnapshot -> {
            versionCode = String.valueOf(documentSnapshot.get("current_version_code"));
            if (versionCode.equalsIgnoreCase(versionCodeApp)) {
                int SPLASH_TIME_OUT = 1500;
                /*
                 * Showing splash screen with a timer. This will be useful when you
                 * want to show case your app logo / company
                 */
                new Handler().postDelayed(() -> {
                    // This method will be executed once the timer is over
                    // Start your app main activity
                    // Session manager
                    PrefManager session = new PrefManager(getApplicationContext());

                    // Check if user is already logged in or not
                    if (session.isFirstTimeLaunch()) {
                        //First time user (Start WelcomeActivity)
                        Intent intent = new Intent(SplashActivity.this, WelcomeActivity.class);
                        startActivity(intent);
                        finish();
                    } else {
                        //User is already logged (Start MainActivity)
                        Intent i = new Intent(SplashActivity.this, MainActivity.class);
                        startActivity(i);
                        finish();
                    }
                }, SPLASH_TIME_OUT);

            }else if (Integer.parseInt(versionCode) < Integer.parseInt(versionCodeApp)){
                //User is on the pre-release version {Start MainActivity}
                /*
                    This case happens only when developer increase the version code of the app
                    and try to use the app for testing
                 */
                Intent intent = new Intent(SplashActivity.this, MainActivity.class);
                startActivity(intent);
                finish();
            }else {
                //User is not on the latest version {Start Update Version Activity}
                Intent intent = new Intent(SplashActivity.this, UpdateVersionActivity.class);
                startActivity(intent);
                finish();
            }
        });
    }
}

File: app/src/main/java/in/completecourse/SplashActivity.kt
package `in`.completecourse

import `in`.completecourse.helper.PrefManager
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import androidx.appcompat.app.AppCompatActivity
import com.google.firebase.firestore.FirebaseFirestore
import java.util.*

class SplashActivity : AppCompatActivity() {
    private val versionCodeApp:String = BuildConfig.VERSION_CODE.toString()
    private var versionCode:String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_splash)

        getVersionCode()

    }

    private fun getVersionCode(){
        FirebaseFirestore.getInstance().collection("flags").document("version_code").get().addOnSuccessListener { documentSnapshot ->
            versionCode = documentSnapshot.get("current_version_code").toString()
            if (versionCode != null) {
                val session = PrefManager(applicationContext)
                when {
                    versionCode.equals(versionCodeApp) -> {
                        val splashTimeOut = 1500

                        /*
                    Showing splash screen with a timer. This will be useful when you
                    want to showcase your app logo/company
                     */
                        /*
                     * Showing splash screen with a timer. This will be useful when you
                     * want to show case your app logo / company
                     */
                        Handler().postDelayed({

                            // This method will be executed once the timer is over
                            // Start your app main activity
                            // Session manager


                            // Check if user is already logged in or not
                            if (session.isFirstTimeLaunch()) {
                                //First time user (Start WelcomeActivity)
                                val intent = Intent(this@SplashActivity, WelcomeActivity::class.java)
                                startActivity(intent)
                                finish()
                            } else {
                                //User is already logged (Start MainActivity)
                                val i = Intent(this@SplashActivity, MainActivity::class.java)
                                startActivity(i)
                                finish()
                            }
                        }, splashTimeOut.toLong())
                    }
                    versionCode!!.toInt() < versionCodeApp.toInt() -> {
                        //User is on the pre-release version {Start WelcomeActivity}
                        /*
                    This case happens only when developer increase the version code of the app
                    and try to use the app for testing or when developer forget to increase the version code in the console but the latest
                    version of the app is live in Google Play
                     */
                        // Check if user is already logged in or not
                        if (session.isFirstTimeLaunch()) {
                            //First time user (Start WelcomeActivity)
                            val intent = Intent(this@SplashActivity, WelcomeActivity::class.java)
                            startActivity(intent)
                            finish()
                        } else {
                            //User is already logged (Start MainActivity)
                            val i = Intent(this@SplashActivity, MainActivity::class.java)
                            startActivity(i)
                            finish()
                        }
                    }
                    else -> {
                        //User is not on the latest version {Start UpdateVersion Activity}
                        val intent = Intent(this, UpdateVersionActivity::class.java)
                        startActivity(intent)
                        finish()
                    }
                }
            }
        }
    }
}